Fix selection box + scrolling in animation editor.

SaracenOne opened the pull request:
Closes #76819

https://github.com/user-attachments/assets/32950993-b599-4386-966b-f01e75fec299

Attempts to fix this by introducing the following changes:
* Disable the panning behaviour when the selection box is active.
* Places the selection box in a clipped container used for the timeline.
* Dynamically repositions the selection box in response to changes to the horizontal and vertical scrollers.

In theory, it should also work with horizontal scrolling too, but I don't have access to a mouse which can do this. If anyone has access to one (I think most Mac computers have a mouse which can do this), please confirm that this fix also works for horizontal scrolling.

SaracenOne committed:
Fix selection box + scrolling in animation editor.

KoBeWi commented:
> In theory, it should also work with horizontal scrolling too, but I don't have access to a mouse which can do this.

You can scroll horizontally by holding Shift.

KoBeWi commented on editor/animation_track_editor.cpp:
```diff
@@ -5724,10 +5728,9 @@ void AnimationTrackEditor::_scroll_input(const Ref<InputEvent> &p_event) {
-		rect = scroll_rect.intersection(rect);
-		box_selection->set_position(rect.position);
-		box_selection->set_size(rect.size);
+		Rect2 visual_rect = Rect2(from - scroll->get_global_position(), rect.get_size());
+		box_selection->set_position(visual_rect.position);
+		box_selection->set_size(visual_rect.size);
```
`visual_rect` doesn't look necessary. You can just directly assign values.

KoBeWi commented on editor/animation_track_editor.h:
```diff
@@ -498,6 +498,10 @@ class AnimationTrackEditor : public VBoxContainer {
 
 	PropertyInfo _find_hint_for_track(int p_idx, NodePath &r_base_path, Variant *r_current_val = nullptr);
 
+	void _scroll_changed(Vector2 p_val);
```
```suggestion
	void _scroll_changed(const Vector2 &p_val);
```

KoBeWi commented on editor/animation_track_editor.cpp:
```diff
@@ -5747,6 +5750,42 @@ void AnimationTrackEditor::_toggle_bezier_edit() {
 
+void AnimationTrackEditor::_scroll_changed(Vector2 p_val) {
+	if (box_selecting) {
+		Vector2 scroll_difference = p_val - prev_scroll_position;
```
```suggestion
		const Vector2 scroll_difference = p_val - prev_scroll_position;
```

KoBeWi commented on editor/animation_track_editor.cpp:
```diff
@@ -5747,6 +5750,42 @@ void AnimationTrackEditor::_toggle_bezier_edit() {
+		}
+
+		Rect2 rect(from, to - from);
+		Rect2 visual_rect = Rect2(from - scroll->get_global_position(), rect.get_size());
```
Again, unnecessary.

KoBeWi commented on editor/animation_track_editor.cpp:
```diff
@@ -7470,8 +7519,8 @@ AnimationTrackEditor::AnimationTrackEditor() {
-	add_child(box_selection);
-	box_selection->set_as_top_level(true);
+	box_selection_container->add_child(box_selection);
+	box_selection->set_as_top_level(false);
```
This is default.

KoBeWi replied:
```suggestion
```
This is default.

KoBeWi commented on editor/animation_track_editor.cpp:
```diff
@@ -7247,16 +7286,26 @@ AnimationTrackEditor::AnimationTrackEditor() {
 
+	scroll->get_v_scroll_bar()->connect(SceneStringName(value_changed), callable_mp(this, &AnimationTrackEditor::_v_scroll_changed));
+	scroll->get_h_scroll_bar()->connect(SceneStringName(value_changed), callable_mp(this, &AnimationTrackEditor::_h_scroll_changed));
+	prev_scroll_position = Vector2(0.0, 0.0);
```
```suggestion
```
Not needed.

SaracenOne commented:
Updated to fix the issues raised. Regarding the Shift+Mouse scroll thing, I believe this is a feature of the ViewPanner which is attached to the Timeline scrollbox, which also gives other navigation controls like being able to scroll through the timeline by clicking and dragging with the middle mouse button. I modified the behaviour to intentionally 'not' call the panner while the selection box is active due to certain undesired behaviour in the site.

Perhaps it might be worth keeping the panner active instead, but modifying its functionality when the selection box is active to disable certain features. Regardless, I still think we need testing with omnidirectional mouse scrolling, but I think it would also still be okay to merge *as is* since its definitely an improvement on the old behaviour in any capacity.

akien-mga commented:
Thanks!
